isthmus/src/main/java/io/substrait/isthmus/expression/ExpressionRexConverter.java

```diff
@@ -248,6 +251,34 @@ public RexNode visit(Expression.ListLiteral expr) throws RuntimeException {
     return rexBuilder.makeCall(SqlStdOperatorTable.ARRAY_VALUE_CONSTRUCTOR, args);
   }
 
+  /**
+   * A custom Calcite array value constructor for empty arrays. The default Calcite {@link
+   * SqlArrayValueConstructor} does not allow for an empty element list, in particular because its
+   * {@code inferReturnType()} method works by inspecting the types of its elements.
+   *
+   * <p>See the <a href="https://issues.apache.org/jira/browse/CALCITE-3504">CALCITE-3504</a> for
+   * more details.
+   */
+  private static class EmptySqlArrayValueConstructor extends SqlArrayValueConstructor {
+    private final RelDataType type;
+
+    private EmptySqlArrayValueConstructor(RelDataType type) {
+      this.type = type;
+    }
+
+    @Override
+    public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
+      return type;
+    }
+  }
+
+  @Override
+  public RexNode visit(Expression.EmptyListLiteral expr) throws RuntimeException {
+    var calciteType = typeConverter.toCalcite(typeFactory, expr.getType());
+    return rexBuilder.makeCall(
+        new EmptySqlArrayValueConstructor(calciteType), Collections.emptyList());
```

Comment:
The docs for `makeCall` mention
```
If you already know the return type of the call, then {@link #makeCall(org.apache.calcite.rel.type.RelDataType, org.apache.calcite.sql.SqlOperator, java.util.List)} is preferred.
 ```

I tried out the following here
```java
    return rexBuilder.makeCall(
        calciteType, new SqlArrayValueConstructor(), Collections.emptyList());
```
here and your test passed. This code bypasses return type inference so it doesn't fail.

Would it make sense to you to construct it directly like this and avoid the need for the `EmptySqlArrayValueConstructor` abstraction?

Reply:
This is a great idea - seems to work, and no need for the `EmptySqlArrayValueConstructor` ✨